kimchi/src/circuits/polynomials/poseidon.rs

```diff
@@ -76,30 +87,28 @@ pub fn constraint<F: FftField + SquareRootField>() -> E<F> {
         })
         .collect();
 
-    for e in ROUND_EQUATIONS.iter() {
-        let &RoundEquation {
-            source,
-            target: (target_row, target_round),
-        } = e;
-        let sboxed: Vec<_> = round_to_cols(source)
-            .map(|i| {
-                cache.cache(
-                    E::cell(Column::Witness(i), Curr).pow(PlonkSpongeConstants15W::SPONGE_BOX),
-                )
-            })
-            .collect();
+    for (round, eq) in ROUND_EQUATIONS.iter().enumerate() {
+        // sbox
+        let mut sboxed = vec![];
+        for input_col in eq.input_cols.clone() {
+            let res =
+                E::cell(Column::Witness(input_col), Curr).pow(PlonkSpongeConstants15W::SPONGE_BOX);
+            sboxed.push(cache.cache(res));
+        }
 
-        res.extend(round_to_cols(target_round).enumerate().map(|(j, col)| {
-            let rc = E::cell(Column::Coefficient(idx), Curr);
+        for (state_i, output_col) in eq.output_cols.clone().enumerate() {
+            // round constant
+            let mut output = E::cell(Column::Coefficient(round * 3 + state_i), Curr);
```

Comment:
If feels a bit strange to hard-code 3 here, but have the ranges be defined separately. I know that they match now, but this will be confusing if we change the number again.

Reply:
Perhaps a `STATE_SIZE` constant instead of a literal 3 here?